

#############Read excel tables################

library(tidyverse)
library(readxl)
library(universalmotif)
rm(list=ls())

source("code/read_excel_tables.R")

# This table contains background subtracted PWMs for all of the TF pairs and individual TFs analyzed.  First line for each factor contains information as follows:	
# Base:	A, C, G or T
# symbol(s):	HGNC symbol (gene names)
# clone type:	Type of construct used: extended DNA binding domain (eDBD) or full-length (FL)
# families:	TF structural family
# experiment type:	Type of experiment: HT-SELEX or methyl-HT-SELEX
# ligand sequence:	Name of primer used to generate selection ligand; number before N indicates length of random sequence
# batch:	batch of HT-SELEX and Methyl-HT-SELEX experiment
# seed:	Sequence used to identify subsequences to be included in the model
# multinomial:	Hamming distance to seed used to identify subsequences to be included in the model
# cycle:	SELEX cycle used to construct the model; previous cycle was used as background unless indicated otherwise by a letter "b" followed by the used background cycle, "u" letter indicates that only the first instance of identical sequencing reads were used in the construction of the PWM
# Matrix is one of the representative PWMs:	value yes or no
# comment:	comment field


all_table <- read_excel("../../PWMs/Yin2017/NIHMS1670407-supplement-Supplemental_tables_S1-6.xlsx", 
                        sheet="S2 PWM models", col_names=FALSE, skip=22, n_max=8996-22, col_types=c(rep("text",12), rep("numeric",19)))

 
PWMs=split_df(all_table)[[1]]

PWMs_metadata=do.call(rbind,lapply(seq(1,nrow(PWMs),5), function(i) PWMs[i,-1]))



colnames(PWMs_metadata)=c("symbol",	"clone","family",	"experiment",
                          "ligand",	"batch", "seed",	"multinomial",
                          "cycle","representative", "comment")
PWMs_metadata$organism="Homo_sapiens"
PWMs_metadata$study="Yin2017"
PWMs_metadata$short=NA
PWMs_metadata$type=NA


#symbol     clone family organism     study      experiment ligand        batch  seed               multinomial cycle representative short type              comment filename         

PWMs_metadata=select(PWMs_metadata,c("symbol",	"clone","family", "organism",	"study","experiment",
                                     "ligand",	"batch", "seed",	"multinomial",
                                     "cycle","representative", "short", "type","comment")
)


#remove / from family names
PWMs_metadata=PWMs_metadata %>% 
  mutate(family = gsub("/|\\.\\d+[A-Za-z]+", "_aka_", family))

PWMs_metadata %>%
  count(symbol) 

PWMs_metadata$filename=NA


PWMs_list=lapply(seq(1,nrow(PWMs),5), function(i) PWMs[(i+1):(i+4),] %>%
                   select_if(~ !any(is.na(.))) )

dir.create(paste0("../../PWMs/Yin2017/pwms/","Homo_sapiens"), recursive=TRUE)
dir.create(paste0("../../PWMs/Yin2017/pwms_space/","Homo_sapiens"), recursive=TRUE)
dir.create(paste0("../../PWMs/Yin2017/transfac/","Homo_sapiens"), recursive=TRUE)
append=FALSE


#Sahu et al 2022

#Known TF motifs, a set of X HT-SELEX motifs were collected (refs 10,29 and 30 and unpublished draft motifs). 
#A more compact set of motifs representing different binding specificities was generated by first constructing a dominating set (880)
#covering motifs from the above sources suing the same method and motif distance threshold as in 29.

#In order to remain information of TF binding differences between methylated and non-methylated DNA ligand (10), 
#for each methyl (or non-methyl) motif in the dominating set, the closest non-methyl (methyl) motif of the same TF was added to the
#representative set (respectively) if it was not yet in the set. This resulted in a representative set of 1121 HT-SELEX motifs (Supplementary Table 3).

#In cases where HT-SELEX motifs for several TFs are highly similar, the motifs have been named in figures according to TF class or subclass,
#in order to highlight that we do not know which of the TFs that have similar motifs binds to the motif in the cells. 
#Same principle has been applied also when specificities of closely related TFs have not been measured.
#Supplementary Table 5 shows the naming for the motifs in each figure.

#Additionally, the following promoter core motifs were collected for TSS analyses from literature: 
#TATA box, Initiator, CCAAT-box, GC-box from ref. 31, BRE, MTE, DPE from ref. 32, and BANP from ref. 33.

#A control set of reversed but not complemented motifs was generated by reversing the column order of each motif matrix


#Methyl minus/plus etc information
#Table S3 - bisulfite-SELEX dinucleotide counts and enrichment	

#"symbol",	"clone","family", "organism",	"study","experiment",
#"ligand",	"batch", "seed",	"multinomial",
#"cycle","representative", "short", "type","comment

#section a: bisulfite SELEX analysis of mCG enrichment	
#TF_name	#HGNC name of TF
#clone Type of clone construct used: extended DNA binding domain (eDBD) or full-length (FL)
#call	#final classification of CG positions 
      #to Little effect, 
      #or MethylPlus 
      #or MethylMinus; 
      #cases where data is inconsistent are labeled as 'inconclusive', 
      #asterisk indicates TFs with different bisulfite-SELEX calls in different experiments. 
      #final call is based on bisulfite call except for cases where bisulfite experiment had low enrichment, low complexity seed, or no data.
#methyl-SELEX_call	call from methyl-SELEX
#bisulfite_call	#call from bisulfite-SELEX. Note also that cut-off was set at 10% and 50% to indicate weak effect and strong effect.
#comment	#comment field
#seed	#seed used to count the dinucleotides (counted dinucleotide position is in bold)
#start_position_of_CG	#position of the counted dinucleotide
#enrichment_of_mCG	#enrichment of methyl-CG (see Methods for details)
#CG_frequency	#frequency of CG at the position indicated
#TG_frequency	#frequency of TG at the position indicated


	
bisulfite_selex <- read_excel("../../PWMs/Yin2017/NIHMS1670407-supplement-Supplemental_tables_S1-6.xlsx", 
                        sheet="S3 bisulfite-SELEX data", col_names=FALSE, skip=21, n_max=543-21,col_types=c(rep("text",7), rep("numeric",10))) #n_max=8996-22, 

#b table, replicate experiments	
bisulfite_selex_replicate <- read_excel("../../PWMs/Yin2017/NIHMS1670407-supplement-Supplemental_tables_S1-6.xlsx", 
                                        sheet="S3 bisulfite-SELEX data", col_names=FALSE, skip=548, n_max=599-548,col_types=c(rep("text",7), rep("numeric",10)))

colnames(bisulfite_selex)=colnames(bisulfite_selex_replicate)=c("symbol",
  "clone",
  "call",
  "methyl-SELEX_call",
  "bisulfite_call", 
  "comment",
  "seed",
  "start_position_of_CG",
  "enrichment_of_mCG",
  "CG_frequency_cycle_3_normal_seq",
  "TG_frequency_cycle_3_normal_seq",
  "CG_frequency_cycle_3_bisulfite_seq",
  "TG_frequency_cycle_3_bisulfite_seq",
  "CG_frequency_cycle_4_normal_seq",
  "TG_frequency_cycle_4_normal_seq",
  "CG_frequency_cycle_4_bisulfite_seq",
  "TG_frequency_cycle_4_bisulfite_seq")
  

#section c: TFs for which bisulfite data was not obtained
#TF_name		call	methyl-SELEX_call		comment	seed for methyl-SELEX

TFs_without_bisulfite_data <- read_excel("../../PWMs/Yin2017/NIHMS1670407-supplement-Supplemental_tables_S1-6.xlsx", 
                                         sheet="S3 bisulfite-SELEX data", col_names=FALSE, skip=605, n_max=703-605,col_types=c(rep("text",7)))

TFs_without_bisulfite_data =TFs_without_bisulfite_data[,-c(2,5)]
colnames(TFs_without_bisulfite_data) <- c("symbol", "call", "methyl-SELEX_call", "bisulfite_call", "comment", "seed_methyl-SELEX")

#All possible outcomes
#Table S5 - Contingency table describing the combined effect of the methyl-preference of a TF, 
#methylation of its site, and biological activity of the TF on the methylated state of the bound CpG

#TFs preference: The binding of TFs to DNA is inhibited (MethylMinus), enhanced (MethylPlus) or not affected (Little effect) by CpG methylation.
#state : The state of the cytosine, non-methylated cytosine (C) or methylated cytosine (mC)
#binding: TFs could (YES) or couldn't (No) bind to subsequences with methylated or non-methylated CpG site(s).
#effect on C:  The effect on cytosine after TFs binding to DNA, 
#inducing methylation on normal cytosine (methylation), 
#demethylation of methylated cytosine (demethylation) 
#or keeping the original state (no effect).

possible_outcomes <- read_excel("../../PWMs/Yin2017/NIHMS1670407-supplement-Supplemental_tables_S1-6.xlsx", 
                                         sheet="S5 All possible outcomes", col_names=FALSE, skip=8, n_max=32-8,col_types=c(rep("text",5)))
colnames(possible_outcomes)=c("TFs_preference", "state", "binding", "effect_on_C", "outcome")

#match metadata and methylation info

as.vector(unique(PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX") %>% select(symbol)))$symbol[1] #541

Methyl_SELEX_metadata <- PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX")

for(TF in unique((PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX") %>% select(symbol))$symbol)[1:5]){
  print("TF")
  print(TF)
  print("Methyl_SELEX")
  print(Methyl_SELEX_metadata[which(Methyl_SELEX_metadata$symbol==TF),])
  print(Methyl_SELEX_metadata[which(Methyl_SELEX_metadata$symbol==TF),"seed"])
  print("bisulfite_SELEX")
  print(bisulfite_selex[which(bisulfite_selex$symbol==TF),])
  print(bisulfite_selex[which(bisulfite_selex$symbol==TF),"seed"])
  print("bisulfite_SELEX_replicate")
  print(bisulfite_selex_replicate[which(bisulfite_selex_replicate$symbol==TF),])
  print(bisulfite_selex_replicate[which(bisulfite_selex_replicate$symbol==TF),"seed"])
  print("TFs_without_bisulfite_data")
  print(TFs_without_bisulfite_data[which(TFs_without_bisulfite_data$symbol==TF),])
#  print(TFs_without_bisulfite_data[which(TFs_without_bisulfite_data$symbol==TF),"seed"])
}

PWMs_metadata[which(PWMs_metadata$symbol==as.vector(unique(PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX") %>% select(symbol)))$symbol[1]), ]
bisulfite_selex[which(bisulfite_selex$symbol==as.vector(unique(PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX") %>% select(symbol)))$symbol[1]), "seed"]

test=PWMs_metadata %>% filter(experiment=="Methyl-HT-SELEX") %>% left_join(bisulfite_selex, by=c("seed"))

flights2 %>% left_join(planes, by = "tailnum")


which(PWMs_metadata$experiment=="Methyl-HT-SELEX")





for(m in 1:length(PWMs_list)){
  
  write.table(PWMs_list[[m]][,-1],row.names = FALSE, col.names=FALSE, quote=FALSE,
              file=paste0("../../PWMs/Yin2017/pwms/","Homo_sapiens","/", 
                          paste0(PWMs_metadata[m,
                          -which(colnames(PWMs_metadata)%in% c("clone", "family","organism", "study","comment", "short", "type", "filename"))], collapse="_"),
                          ".pfm"), sep="\t")
  
  file=paste0("../../PWMs/Yin2017/pwms/","Homo_sapiens","/", 
              paste0(PWMs_metadata[m,
                                   -which(colnames(PWMs_metadata)%in% c("clone", "family","organism", "study","comment", "short", "type", "filename"))], collapse="_"),
              ".pfm")
  
  if ( ncol(PWMs_list[[m]][,-1])!=0 ){
    motif=universalmotif::read_matrix(file=file, sep="\t", header=FALSE)
  
    motif@name=paste0(PWMs_metadata[m,-which(colnames(PWMs_metadata)%in% c("clone", "family", "organism", "study","comment", "short", "type", "filename"))], collapse="_")
  
    transfac=paste0("../../PWMs/Yin2017/transfac/",PWMs_metadata[m,"organism"],"/", 
                  paste0(PWMs_metadata[m,-which(colnames(PWMs_metadata)%in% c("clone","family","comment", "study","organism","short", "type","filename"))], collapse="_"),".pfm")
    write_transfac(motif, file=transfac, overwrite = TRUE, append = FALSE)
    
  }
  
  
  write.table(PWMs_list[[m]][,-1],row.names = FALSE, col.names=FALSE, quote=FALSE,
              file=paste0("../../PWMs/Yin2017/pwms_space/","Homo_sapiens","/", 
                          paste0(PWMs_metadata[m,
                                               -which(colnames(PWMs_metadata)%in% c("clone", "family","organism", "study","comment", "short", "type", "filename"))], collapse="_"),
                          ".pfm"), sep=" ")
  
  PWM=as.matrix(PWMs_list[[m]][,-1], dimnames=NULL)
  rownames(PWM)=c("A", "C", "G", "T")
  write.table(paste0(">",  paste0(PWMs_metadata[m,-which(colnames(PWMs_metadata)%in% c("clone", "family", "organism", "study","comment", "short", "type", "filename"))], collapse="_")),   
              append=append, row.names = FALSE, col.names=FALSE, quote=FALSE,
              file=paste0("../../PWMs/Yin2017/",PWMs_metadata[m,"organism"],"_all", ".scpd"))
  append=TRUE
  
  write.table(PWM,append=append, row.names = TRUE, col.names=FALSE, quote=FALSE,
              file=paste0("../../PWMs/Yin2017/",PWMs_metadata[m,"organism"],"_all", ".scpd"))
  
  PWMs_metadata$filename[m]=paste0("PWMs/Yin2017/pwms/Homo_sapiens/", paste0(PWMs_metadata[m,
                                              -which(colnames(PWMs_metadata)%in% c("clone", "family","organism", "study","comment", "short", "type", "filename"))], collapse="_"),".pfm")
  
}

write.table(PWMs_metadata, file="../../PWMs/Yin2017/metadata.csv", row.names = FALSE,sep="\t")
saveRDS(PWMs_metadata, file="data/Yin2017.Rds")




#display_table_shape(all_table)

