#install.packages("ggseqlogo")
library("ggseqlogo")
library("motifStack")
library("ggplot2")
#packageVersion("motifStack") #1.42.0
#A control set of reversed but not complemented motifs was generated by reversing the column order of each motif matrix

#Promoter core motifs: TATA box, Initiator, CCAAT-box, GC-box, BRE, MTE, DEP, BANP Ref 31,32,33

#setwd("/projappl/project_2006203/TFBS/RProjects/TFBS/Experiments")

representatives=read.table("/scratch/project_2006203/motif-clustering-Viestra-private/metadata/new_representatives_IC_length_Morgunova.tsv", sep="\t", header=TRUE)

#make directories

data_path="/scratch/project_2006203/TFBS"

dir.create(file.path("/scratch/project_2006203/TFBS", "Logos"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos", "pdf"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos", "png"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/pdf", "barcode"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/pdf", "ic"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/pdf", "prob"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/png", "barcode"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/png", "ic"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos/png", "prob"), showWarnings = FALSE)

dir.create(file.path("/scratch/project_2006203/TFBS", "Logos2"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2", "pdf"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2", "png"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/pdf", "barcode"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/pdf", "ic"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/pdf", "prob"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/png", "barcode"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/png", "ic"), showWarnings = FALSE)
dir.create(file.path("/scratch/project_2006203/TFBS/Logos2/png", "prob"), showWarnings = FALSE)





pcm_list<-TFBSTools::PFMatrixList()
pwm_list<-TFBSTools::PWMatrixList()

for(i in 1:nrow(representatives)){
  print(i)
  pcm=as.matrix( read.table(paste0(data_path,"/", representatives$filename[i]), header=FALSE) )
  dimnames(pcm)=list(c("A", "C", "G", "T"))
  #A 1294  802 6919 6919    0  156 6919 1801
  #C 6919 3055 1044  249   41 1461  493 2373
  #G 3132  905  938   56    0  536 1133 1625
  #T 2804 3864  657    0 6919 6919  462 1120
  
  
  pcm_class <- TFBSTools::PFMatrix(ID=representatives$ID[i], name=representatives$symbol[i], 
                                   #matrixClass="Zipper-Type", 
                                   strand="+", 
                                   bg=c(A=0.25, C=0.25, G=0.25, T=0.25), 
                                   tags=list(family=representatives$Lambert2018.families[i], 
                                             species=representatives$organism[i] 
                                             #tax_group="vertebrates", 
                                             #medline="7592839", 
                                             #type="SELEX", 
                                             #ACC="P53762", 
                                             #pazar_tf_id="TF0000003",
                                             #TFBSshape_ID="11", 
                                             #TFencyclopedia_ID="580"
                                   ),
                                   profileMatrix=pcm
  )
  
  
  
  
  pwm_class <- TFBSTools::toPWM(pcm_class, type="prob", pseudocounts = 0.01, bg=c(A=0.25, C=0.25, G=0.25, T=0.25))
  
  
  
  pcm_list[[representatives$ID[i]]]=pcm_class
  pwm_list[[representatives$ID[i]]]=pwm_class
  
  #pwm_class@profileMatrix
  #A 0.09145534 0.09297491 0.72389572 9.577786e-01 3.591949e-07 0.01719602 0.76817973 0.2602977
  #C 0.48900966 0.35416172 0.10922802 3.446874e-02 5.891155e-03 0.16104507 0.05473542 0.3429685
  #G 0.22135842 0.10491554 0.09813784 7.752273e-03 3.591949e-07 0.05908310 0.12579119 0.2348606
  #T 0.19817659 0.44794783 0.06873842 3.460682e-07 9.941081e-01 0.76267580 0.05129366 0.1618732
  
  #pwm=motifStack::pcm2pfm(pcm)
  #A 0.09145523 0.1542689 0.5701286 0.6113683 0.1270231 0.1007315 0.5798643 0.2550357
  #C 0.48900982 0.3135027 0.1549049 0.1399569 0.1299208 0.1929642 0.1256979 0.2954626
  #G 0.22135840 0.1615485 0.1474132 0.1263163 0.1270231 0.1275885 0.1709308 0.2425966
  #T 0.19817655 0.3706799 0.1275532 0.1223585 0.6160329 0.5787158 0.1235070 0.2069051
  
  #seqLogo_pwms=lapply( lapply(example_pwms, function(x) x@profileMatrix), seqLogo::makePWM)
  
  #motif <- new("pfm", mat=pwm_class@profileMatrix, name=pwm_class@name)
  
  #motifStack::plotMotifLogo( pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = FALSE, ylab="probability", font="mono,Courier", fontface = "plain")
  #motifStack::plotMotifLogo(pwm, ic.scale = FALSE) Something goes wrong
  
  #motifStack::plotMotifLogo(pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = TRUE, font="mono,Courier", fontface = "light")
  
  #Probability matrices
  png(paste0(data_path,"/Logos2/png/prob/",representatives$ID[i],".png"), res=600,  width = 2500/4*representatives$length[i], height = 2500)
  print(motifStack::plotMotifLogo( pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = FALSE, 
                             ylab="probability", font="mono,Courier", fontface = "bold"))
  dev.off()
  
  pdf(paste0(data_path,"/Logos2/pdf/prob/",representatives$ID[i],".pdf"), width =representatives$length[i], height = 4)
  print(motifStack::plotMotifLogo( pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = FALSE, 
                             ylab="probability", font="mono,Courier", fontface = "bold"))
  dev.off()
  
  png(paste0(data_path,"/Logos/png/prob/",representatives$ID[i],".png"), res=600,  width = 2500/4*representatives$length[i], height = 2500)
  print(ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="probability", font="roboto_medium", col_scheme="auto"  ) + 
    ggseqlogo::theme_logo()  +
    labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none"))
  
  dev.off()
  
  
  pdf(paste0(data_path,"/Logos/pdf/prob/",representatives$ID[i],".pdf"), width =representatives$length[i], height = 4)
  print(ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="probability", font="roboto_medium", col_scheme="auto" ) + 
    ggseqlogo::theme_logo()  +
    labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none"))
  dev.off()
  
  #Information content matrices
  
  
  png(paste0(data_path,"/Logos2/png/ic/",representatives$ID[i],".png"), res=600,  width = 2500/4*representatives$length[i], height = 2500)
  print(motifStack::plotMotifLogo( pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = TRUE, 
                             ylab="bits", font="mono,Courier", fontface = "bold"))
  dev.off()
  
  pdf(paste0(data_path,"/Logos2/pdf/ic/",representatives$ID[i],".pdf"), width =representatives$length[i], height = 4)
  print(motifStack::plotMotifLogo( pwm_class@profileMatrix, motifName=pwm_class@name, ic.scale = TRUE, 
                             ylab="bits", font="mono,Courier", fontface = "bold"))
  dev.off()
  
  png(paste0(data_path,"/Logos/png/ic/",representatives$ID[i],".png"), res=600,  width = 2500/4*representatives$length[i], height = 2500)
  print(ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="bits", font="roboto_medium", col_scheme="auto"  ) + 
    ggseqlogo::theme_logo()  +
    labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none"))
  
  dev.off()
  
  
  pdf(paste0(data_path,"/Logos/pdf/ic/",representatives$ID[i],".pdf"), width =representatives$length[i], height = 4)
  print(ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="bits", font="roboto_medium", col_scheme="auto" ) + 
    ggseqlogo::theme_logo()  +
    labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none"))
  dev.off()
  
  
  
}

saveRDS(pcm_list, file=paste0("/scratch/project_2006203/TFBS/RProjects/TFBS/RData/pcm_list.Rds")
saveRDS(pwm_list, file="/scratch/project_2006203/TFBS/RProjects/TFBS/RData/pwm_list.Rds")


#change the x-axis labels
#plot(motif, xaxis=paste0("pos", seq.int(7)+10))

#plot sequence logo stack
#To show multiple motifs on the same canvas as a sequence logo stack, 
#the distance of motifs need to be calculated first. 
#Previously, MotIV3::motifDistances ( R implementation of STAMP4) is 
#used to calculate the distance. However, 
#The MotIV package were dropped from Bioconductor 3_12. 
#Currently, by default, R implementation of matalign is used. 
#After alignment, users can use plotMotifLogoStack, 
#plotMotifLogoStackWithTree or plotMotifStackWithRadialPhylog 
#to draw sequence logos in different layouts. 
#To make it easy to use, we integrated different functionalities into one workflow function named as motifStack.

#motifStack::motifStack(motifs, layout="stack", ncex=1.0, reorder=FALSE, layout="tree")

#Affinity logos

# plot a sequence logo cloud
# We can also plot a sequence logo cloud for DNA motifs.
# 
# ## assign groups for motifs
# groups <- rep(paste("group",1:5,sep=""), each=10)
# names(groups) <- names(pfms)
# ## assign group colors
# group.col <- brewer.pal(5, "Set3")
# names(group.col)<-paste("group",1:5,sep="")
# ## create a list of pfm objects
# pfms <- mapply(names(pfms), pfms, FUN=function(.ele, .pfm){
#   new("pfm",mat=.pfm, name=.ele)}
#   ,SIMPLIFY = FALSE)
# ## use matalign to calculate the distances of motifs
# hc <- clusterMotifs(pfms)
# ## convert the hclust to phylog object
# library(ade4)
# phylog <- ade4::hclust2phylog(hc)
# ## reorder the pfms by the order of hclust
# leaves <- names(phylog$leaves)
# pfms <- pfms[leaves]
# ## extract the motif signatures
# motifSig <- motifSignature(pfms, phylog, cutoffPval=0.0001, min.freq=1)
# ## draw the motifs with a tag-cloud style.
# motifCloud(motifSig, scale=c(6, .5), 
#            layout="rectangles", 
#            group.col=group.col, 
#            groups=groups, 
#            draw.legend=TRUE)

# representatives$ID[i]


#pdf(paste0("/scratch/project_2006203/TFBS/Logos/prob/",name,".pdf"), res=600,  width = 4000, height = 2500)
# png(paste0("../../Logos/png/prob/",representatives$ID[i],".png"), res=600,  width = 4000, height = 2500)
# p=ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="probability" ) + ggseqlogo::theme_logo()  +
#   labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none")
# plot(p)
# dev.off()
# 
# pdf(paste0("../../Logos/pdf/prob/",representatives$ID[i],".pdf"), width = 15, height = 5)
# 
# p=ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="probability", font="roboto_slab_bold", col_scheme="auto" ) + ggseqlogo::theme_logo()  +
#   labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none")
# plot(p)
# dev.off
# 
# png(paste0("../../Logos/png/prob/",representatives$ID[i],".png"), res=600,  width = 4000, height = 2500)
# p=ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="bits" ) + ggseqlogo::theme_logo()  +
#   labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none")
# plot(p)
# dev.off()
# 
# pdf(paste0("../../Logos/pdf/prob/",representatives$ID[i],".pdf"), width = 15, height = 5)
# 
# p=ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix, method="probability", font="roboto_slab_bold", col_scheme="auto" ) + ggseqlogo::theme_logo()  +
#   labs(title=pwm_class@name)+ theme(title =element_text(size=8, face='bold'))+ guides(scale="none")
# plot(p)
# dev.off()


# ggseqlogo::list_fonts()
# Available ggseqlogo fonts:
#   helvetica_regular
# helvetica_bold
# helvetica_light
# roboto_medium
# roboto_bold
# roboto_regular
# akrobat_bold
# akrobat_regular
# roboto_slab_bold
# roboto_slab_regular
# roboto_slab_light
# xkcd_regular
# 
# ggseqlogo::list_col_schemes()
# auto
# chemistry
# chemistry2
# hydrophobicity
# nucleotide
# nucleotide2
# base_pairing
# clustalx
# taylor


# font="mono,Courier", fontface="plain"
# 
# png(paste0("../../Logos/png/IC/",name,".png"), res=600,  width = 4000, height = 2500)
# p=ggplot() + ggseqlogo::geom_logo(  example_pwms[[name]]@profileMatrix, method="bits" ) + ggseqlogo::theme_logo()  +
#   labs(title=name)+ theme(title =element_text(size=8, face='bold'))
# plot(p)
# #seqLogo(seqLogo_pwms[[name]], fill=c(A="darkgreen", C="blue", G="orange", T="red"))
# dev.off()
# 
# 
# 
# 
# seqLogo::makePWM(pwm_class@profileMatrix)
# 
# install
#ggplot() + ggseqlogo::geom_logo(  pwm_class@profileMatrix ) + ggseqlogo::theme_logo()